Bug Report: `vtorc` panics in `EmergencyReparentShard` when all candidates have errant GTIDs

### Overview of the Issue

If every replica candidate is filtered out by errant GTID detection during `EmergencyReparentShard` (ERS), `reparentShardLocked` passes an empty candidate map to `findMostAdvanced`, which then panics with `runtime error: index out of range [0] with length 0`. Because ERS is invoked from `vtorc`, the panic crashes the `vtorc` process (exit code 2).

Fix: [#19904](https://github.com/vitessio/vitess/pull/19904)

## How it was detected

This was surfaced by an Antithesis run (`2026-04-01 Vitess Test`, ~574 test hours). The failing properties correlated almost 1:1:

| Property | Group | Failing | Passing |
|---|---|---|---|
| `runtime error: index out of range [{*}] with length {*}` | Correctness / No Panics | 520 | 0 |
| `container: vtorc, exit code: 2` | Correctness / No unexpected container exits | 519 | 0 |
| `Vitess cluster health did not recover within timeout` | Antithesis SDK / Unreachability assertions | 171 | 0 |
| `crud/eventually_vitess_health` | Test Composer | 171 | 21,421 |

The near-identical counts (520 panics ↔ 519 `vtorc` exits) show the panic is the direct cause of the container death. The downstream health and CRUD failures (`primaries=0, replicas=2` on `test_keyspace/-80`) are a consequence: once `vtorc` is dead, ERS cannot complete, so no primary is ever promoted and the cluster fails to converge.

### Reproduction Steps

From the Antithesis logs:

At vtime `669.302s`–`669.309s` for `test_keyspace/-80`, the sequence was:

1. `vtorc` starts `PrimarySemiSyncBlocked` recovery and calls `findErrantGTIDs` against the two available replicas (uid `102` and `104`).
2. Both are skipped because they have non-overlapping extra GTIDs:
   - `skipping test-0000000102 ... because we detected errant GTIDs - 0160a321-...:11`
   - `skipping test-0000000104 ... because we detected errant GTIDs - 01669473-...:1-18`
3. `findErrantGTIDs` returns an empty map.
4. `reparentShardLocked` calls `findMostAdvanced` with the empty map.
5. `findMostAdvanced` panics:

```
panic: runtime error: index out of range [0] with length 0

goroutine 8233 [running]:
vitess.io/vitess/go/vt/vtctl/reparentutil.(*EmergencyReparenter).findMostAdvanced(...)
    go/vt/vtctl/reparentutil/emergency_reparenter.go:419 +0x8cf   # validTablets[0]
```

(Line numbers in the binary stack differ slightly from current `main`; in the current source the offending access is `validTablets[0]` / `tabletPositions[0]` in `findMostAdvanced`.)

## Root cause

`go/vt/vtctl/reparentutil/emergency_reparenter.go` — `reparentShardLocked`:

```go
// Restrict the valid candidates list. ...
validCandidates, err = restrictValidCandidates(validCandidates, tabletMap)
if err != nil {
    return err
} else if len(validCandidates) == 0 {
    return vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no valid candidates for emergency reparent")
}

// ...

// For GTID based replication, we will run errant GTID detection.
if isGTIDBased {
    validCandidates, err = erp.findErrantGTIDs(ctx, validCandidates, ...)
    if err != nil {
        return err
    }
    // <-- no empty-map guard here; if findErrantGTIDs filtered everyone out,
    //     findMostAdvanced is called with an empty map and panics.
}

intermediateSource, validCandidateTablets, err = erp.findMostAdvanced(validCandidates, tabletMap, opts)
```

`findMostAdvanced` assumes at least one candidate:

```go
winningPrimaryTablet := validTablets[0]   // panics when validCandidates is empty
winningPosition := tabletPositions[0]
```

The sibling call site (after `restrictValidCandidates`) already has the empty-check; the `findErrantGTIDs` call site was missing the symmetric guard.

## When this is hit

Any scenario where errant GTID detection filters out **every** candidate. In the Antithesis run, this happened after fault injection left the two surviving replicas with mutually exclusive extra GTIDs from their respective prior primary UUIDs.

## Fix

[#19904](https://github.com/vitessio/vitess/pull/19904) adds the missing guard after `findErrantGTIDs`, returning a clear `FAILED_PRECONDITION` error instead of panicking, plus a regression test that reproduces the exact panic (and verifies the new error after the fix).

## Related

- [#19848](https://github.com/vitessio/vitess/pull/19848) previously fixed a different nil-pointer bug *inside* `findErrantGTIDs`, but did not add the empty-map guard after it returns.

### Binary Version

```sh
vtgate version Version: 25.0.0-SNAPSHOT (Git revision c5a8b067310a01d04437674624184454698dd249 branch 'main') built on Thu Apr 16 18:55:26 UTC 2026 by matt@pslord.local using go1.26.2 darwin/arm64
```

### Operating System and Environment details

```sh
N/A
```

### Log Fragments

```sh

```

--- a/go/vt/vtctl/reparentutil/emergency_reparenter.go
+++ b/go/vt/vtctl/reparentutil/emergency_reparenter.go
@@ -489,6 +489,9 @@ func (erp *EmergencyReparenter) findMostAdvanced(
 	opts EmergencyReparentOptions,
 ) (*topodatapb.Tablet, []*topodatapb.Tablet, error) {
 	erp.logger.Infof("started finding the intermediate source")
+	if len(validCandidates) == 0 {
+		return nil, nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no valid candidates for emergency reparent")
+	}
 	// convert the valid candidates into a list so that we can use it for sorting
 	validTablets, tabletPositions, err := getValidCandidatesAndPositionsAsList(validCandidates, tabletMap)
 	if err != nil {

--- a/go/vt/vtctl/reparentutil/emergency_reparenter.go
+++ b/go/vt/vtctl/reparentutil/emergency_reparenter.go
@@ -281,6 +281,9 @@ func (erp *EmergencyReparenter) reparentShardLocked(ctx context.Context, ev *eve
 		if err != nil {
 			return err
 		}
+		if len(validCandidates) == 0 {
+			return vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "no valid candidates for emergency reparent: all candidates have errant GTIDs")
+		}
 	}
 
 	// Find the intermediate source for replication that we want other tablets to replicate from.

--- a/go/vt/vtctl/reparentutil/emergency_reparenter_test.go
+++ b/go/vt/vtctl/reparentutil/emergency_reparenter_test.go
@@ -3194,6 +3194,11 @@ func TestEmergencyReparenter_findMostAdvanced(t *testing.T) {
 		err                  string
 	}{
 		{
+			name:            "no valid candidates",
+			validCandidates: map[string]*RelayLogPositions{},
+			tabletMap:       map[string]*topo.TabletInfo{},
+			err:             "no valid candidates for emergency reparent",
+		}, {
 			name: "choose most advanced",
 			validCandidates: map[string]*RelayLogPositions{
 				"zone1-0000000100": positionMostAdvanced,

--- a/go/vt/vtctl/reparentutil/emergency_reparenter_test.go
+++ b/go/vt/vtctl/reparentutil/emergency_reparenter_test.go
@@ -1869,6 +1869,80 @@ func TestEmergencyReparenter_reparentShardLocked(t *testing.T) {
 			shouldErr:        true,
 			errShouldContain: "primary zone1-0000000100 is not equal to expected alias zone1-0000000101",
 		},
+		{
+			// Regression test: if every candidate has mutually errant GTIDs, findErrantGTIDs
+			// returns an empty map, which previously caused findMostAdvanced to panic with
+			// "index out of range [0] with length 0" when indexing the empty tablet slice.
+			name:                 "all candidates filtered out by errant GTID detection",
+			durability:           policy.DurabilityNone,
+			emergencyReparentOps: EmergencyReparentOptions{},
+			tmc: &testutil.TabletManagerClient{
+				ReadReparentJournalInfoResults: map[string]int32{
+					"zone1-0000000100": 3,
+					"zone1-0000000101": 3,
+				},
+				StopReplicationAndGetStatusResults: map[string]struct {
+					StopStatus *replicationdatapb.StopReplicationStatus
+					Error      error
+				}{
+					"zone1-0000000100": {
+						StopStatus: &replicationdatapb.StopReplicationStatus{
+							Before: &replicationdatapb.Status{IoState: int32(replication.ReplicationStateRunning), SqlState: int32(replication.ReplicationStateRunning)},
+							After: &replicationdatapb.Status{
+								SourceUuid:       "00000000-0000-0000-0000-000000000001",
+								RelayLogPosition: getRelayLogPosition("1-100", "1-31", "1-50"),
+							},
+						},
+					},
+					"zone1-0000000101": {
+						StopStatus: &replicationdatapb.StopReplicationStatus{
+							Before: &replicationdatapb.Status{IoState: int32(replication.ReplicationStateRunning), SqlState: int32(replication.ReplicationStateRunning)},
+							After: &replicationdatapb.Status{
+								SourceUuid:       "00000000-0000-0000-0000-000000000001",
+								RelayLogPosition: getRelayLogPosition("1-100", "1-30", "1-51"),
+							},
+						},
+					},
+				},
+				WaitForPositionResults: map[string]map[string]error{
+					"zone1-0000000100": {
+						getRelayLogPosition("1-100", "1-31", "1-50"): nil,
+					},
+					"zone1-0000000101": {
+						getRelayLogPosition("1-100", "1-30", "1-51"): nil,
+					},
+				},
+			},
+			shards: []*vtctldatapb.Shard{
+				{
+					Keyspace: "testkeyspace",
+					Name:     "-",
+				},
+			},
+			tablets: []*topodatapb.Tablet{
+				{
+					Alias: &topodatapb.TabletAlias{
+						Cell: "zone1",
+						Uid:  100,
+					},
+					Keyspace: "testkeyspace",
+					Shard:    "-",
+				},
+				{
+					Alias: &topodatapb.TabletAlias{
+						Cell: "zone1",
+						Uid:  101,
+					},
+					Keyspace: "testkeyspace",
+					Shard:    "-",
+				},
+			},
+			keyspace:         "testkeyspace",
+			shard:            "-",
+			cells:            []string{"zone1"},
+			shouldErr:        true,
+			errShouldContain: "no valid candidates for emergency reparent",
+		},
 	}
 
 	for _, tt := range tests {

